File: a_star_smooth_planner/a_star_smoother.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from nav_msgs.msg import OccupancyGrid, Path
from geometry_msgs.msg import PoseStamped, Pose
from rclpy.qos import QoSProfile, DurabilityPolicy

from queue import Queue

class GridPose:
    def __init__(self, x=0, y=0):
        self.x = x
        self.y = y



class AStarSmoother(Node):
    def __init__(self):
        super().__init__("a_star_smoother")

        self.declare_parameter("iterations", 2)
        self.declare_parameter("cost_limit", 20)
        self.iterations = self.get_parameter("iterations").value
        self.cost_limit = self.get_parameter("cost_limit").value

        map_qos = QoSProfile(depth=10)
        map_qos.durability = DurabilityPolicy.TRANSIENT_LOCAL

        self.map_sub = self.create_subscription(
            OccupancyGrid, "/costmap", self.map_callback, map_qos
        )
        self.path_sub = self.create_subscription(Path, "/a_star/path", self.path_callback, 10)
        self.path_pub = self.create_publisher(Path, "/a_star/path/smooth", 10)

        self.map_ = None
        self.path_ = None

        self.get_logger().info("a_star_smoother node has just started")

    def pose_to_grid(self, pose: Pose) -> GridPose:
        grid_x = int((pose.position.x - self.map_.info.origin.position.x) / self.map_.info.resolution)
        grid_y = int((pose.position.y - self.map_.info.origin.position.y) / self.map_.info.resolution)
        return GridPose(grid_x, grid_y)

    def grid_to_pose(self, grid_pose: GridPose) -> Pose:
        pose = Pose()
        pose.position.x = grid_pose.x * self.map_.info.resolution + self.map_.info.origin.position.x
        pose.position.y = grid_pose.y * self.map_.info.resolution + self.map_.info.origin.position.y
        return pose

    def grid_to_cell(self, grid_pose: GridPose):
        return grid_pose.y * self.map_.info.width + grid_pose.x
    
    def line_crosses_obstacle(self, line): 
        map_data_list = self.map_.data
        grid_to_cell = self.grid_to_cell
        cost_limit = self.cost_limit       
        for grid_pose in line:
            cell_index = grid_to_cell(grid_pose)
            if map_data_list[cell_index] > cost_limit:
                return True
        return False
    
    def map_callback(self, map_msg: OccupancyGrid):
        self.map_ = map_msg

    def path_callback(self, path_msg: Path):
        if (self.map_ == None):
            self.get_logger().error("No Map data received")
        
        self.get_logger().info("Path received")
        self.path_ = path_msg
        new_path = self.smoothen_path(self.path_)
        
        self.path_pub.publish(new_path)
        self.get_logger().info('smoothned a_star path published')

    def smoothen_path(self, path: Path) -> Path:
        pose_to_grid = self.pose_to_grid
        grid_to_pose = self.grid_to_pose
        # bresenham = self.bresenham_line
        smoothen_grid_path = self.smoothen_grid_path

        grid_pose_list = [pose_to_grid(pose.pose) for pose in path.poses]

        for _ in range(self.iterations):
            grid_pose_list = smoothen_grid_path(grid_pose_list)

        new_path = Path()
        new_path.header.frame_id = self.map_.header.frame_id
        for grid_pose in grid_pose_list:
            pose_stamped = PoseStamped()
            pose_stamped.header.frame_id = self.map_.header.frame_id
            pose_stamped.pose = grid_to_pose(grid_pose)
            new_path.poses.append(pose_stamped)
        
        return new_path
    
    def smoothen_grid_path(self, grid_pose_list):
        pose_queue_check = Queue(maxsize=2)
        new_grid_pose_list = []

        start_grid_pose = grid_pose_list[0]
        goal_grid_pose = grid_pose_list[-1]

        bresenham = self.bresenham_line
        line_crosses_obstacle = self.line_crosses_obstacle

        no_of_path = len(grid_pose_list)
        touch_obstacle = False
        i = 1

        if no_of_path == 1:
            self.get_logger().warning("only one path was found")
            new_grid_pose_list.append(start_grid_pose)
            return 
        
        pose_queue_check.put(start_grid_pose)
        new_grid_pose_list.append(start_grid_pose)
        
        while i < no_of_path and rclpy.ok():
            next_grid_pose = grid_pose_list[i]
            pose_queue_check.put(next_grid_pose)

            line = bresenham(start_grid_pose, next_grid_pose)
            touch_obstacle = line_crosses_obstacle(line)

            if touch_obstacle:
                next_grid_pose = pose_queue_check.get()
                new_grid_pose_list.append(next_grid_pose)
                start_grid_pose = next_grid_pose
                pose_queue_check.get()
                pose_queue_check.put(next_grid_pose)
                touch_obstacle = False
            
            else:
                pose_queue_check.get()
                i+=1

        new_grid_pose_list.append(goal_grid_pose)

        full_new_grid_pose_list = []
        full_new_grid_pose_list.append(new_grid_pose_list[0])
        length = len(new_grid_pose_list)
        for i in range(1, length):
            line = bresenham(new_grid_pose_list[i-1], new_grid_pose_list[i])
            full_new_grid_pose_list.extend(line[1:])
        return full_new_grid_pose_list
    
    def bresenham_line(self, start: GridPose, end: GridPose):
        line = []

        x0, y0, x1, y1 = start.x, start.y, end.x, end.y

        dx = x1 - x0
        dy = y1 - y0

        xsign = 1 if dx > 0 else -1
        ysign = 1 if dy > 0 else -1

        dx = abs(dx)
        dy = abs(dy)

        if dx > dy:
            xx = xsign
            xy = 0
            yx = 0
            yy = ysign
        else:
            tmp = dx
            dx = dy
            dy = tmp
            xx = 0
            xy = ysign
            yx = xsign
            yy = 0

        D = 2 * dy - dx
        y = 0

        for i in range(dx + 1):
            line.append(GridPose(x0 + i * xx + y * yx, y0 + i * xy + y * yy))
            if D >= 0:
                y += 1
                D -= 2 * dx
            D += 2 * dy

        return line



def main(args=None):
    rclpy.init(args=args)
    node = AStarSmoother()
    rclpy.spin(node)
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/a_star_smoother.cpp
#include "rclcpp/rclcpp.hpp"

#include "nav_msgs/msg/occupancy_grid.hpp"
#include "nav_msgs/msg/path.hpp"
#include "geometry_msgs/msg/pose_stamped.hpp"

#include <queue>
#include <vector>
#include <cmath>

using std::placeholders::_1;

struct GridPose
{
  int x{0};
  int y{0};

  GridPose() = default;
  GridPose(int x_, int y_) : x(x_), y(y_) {}
};

class AStarSmoother : public rclcpp::Node
{
public:
  AStarSmoother()
  : Node("a_star_smoother")
  {
    declare_parameter<int>("iterations", 2);
    declare_parameter<int>("cost_limit", 20);

    iterations_  = get_parameter("iterations").as_int();
    cost_limit_  = get_parameter("cost_limit").as_int();

    rclcpp::QoS map_qos(10);
    map_qos.transient_local();

    map_sub_ = create_subscription<nav_msgs::msg::OccupancyGrid>(
      "/costmap", map_qos,
      std::bind(&AStarSmoother::mapCallback, this, _1));

    path_sub_ = create_subscription<nav_msgs::msg::Path>(
      "/a_star/path", 10,
      std::bind(&AStarSmoother::pathCallback, this, _1));

    path_pub_ = create_publisher<nav_msgs::msg::Path>(
      "/a_star/path/smooth", 10);

    RCLCPP_INFO(get_logger(), "A* LOS smoother started");
  }

private:
  /* ===================== Callbacks ===================== */

  void mapCallback(const nav_msgs::msg::OccupancyGrid::SharedPtr msg)
  {
    map_ = msg;
  }

  void pathCallback(const nav_msgs::msg::Path::SharedPtr msg)
  {
    if (!map_) {
      RCLCPP_ERROR(get_logger(), "No costmap received");
      return;
    }

    startTime = std::chrono::system_clock::now();
    
    auto smoothed = smoothenPath(*msg);
    path_pub_->publish(smoothed);

    endTime = std::chrono::system_clock::now();

    testDuration = endTime - startTime;

    RCLCPP_INFO_STREAM(get_logger(), "Smoothed path published in " << testDuration.count() << " sec" );

    // RCLCPP_INFO(get_logger(), "Smoothed path published");
  }

  /* ===================== Core Logic ===================== */

  nav_msgs::msg::Path smoothenPath(const nav_msgs::msg::Path & path)
  {
    std::vector<GridPose> grid_path;
    grid_path.reserve(path.poses.size());

    for (const auto & p : path.poses) {
      grid_path.push_back(poseToGrid(p.pose));
    }

    for (int i = 0; i < iterations_; ++i) {
      grid_path = smoothenGridPath(grid_path);
    }

    nav_msgs::msg::Path out;
    out.header.frame_id = map_->header.frame_id;

    for (const auto & gp : grid_path) {
      geometry_msgs::msg::PoseStamped ps;
      ps.header.frame_id = out.header.frame_id;
      ps.pose = gridToPose(gp);
      out.poses.push_back(ps);
    }

    return out;
  }

  std::vector<GridPose> smoothenGridPath(const std::vector<GridPose> & path)
  {
    if (path.size() <= 1) {
      return path;
    }

    std::queue<GridPose> pose_queue;
    std::vector<GridPose> sparse_path;

    GridPose start = path.front();
    GridPose goal  = path.back();

    pose_queue.push(start);
    sparse_path.push_back(start);

    size_t i = 1;

    while (i < path.size()) {
      GridPose next = path[i];
      pose_queue.push(next);

      auto line = bresenhamLine(start, next);

      if (lineCrossesObstacle(line)) {
        GridPose last_safe = pose_queue.front();
        pose_queue.pop();

        sparse_path.push_back(last_safe);
        start = last_safe;

        while (!pose_queue.empty()) {
          pose_queue.pop();
        }
        pose_queue.push(start);
      } else {
        pose_queue.pop();
        ++i;
      }
    }

    sparse_path.push_back(goal);

    /* Re-densify */
    std::vector<GridPose> dense_path;
    dense_path.push_back(sparse_path.front());

    for (size_t k = 1; k < sparse_path.size(); ++k) {
      auto line = bresenhamLine(sparse_path[k - 1], sparse_path[k]);
      dense_path.insert(dense_path.end(), line.begin() + 1, line.end());
    }

    return dense_path;
  }

  /* ===================== Geometry ===================== */

  GridPose poseToGrid(const geometry_msgs::msg::Pose & pose) const
  {
    int gx = static_cast<int>(
      (pose.position.x - map_->info.origin.position.x) / map_->info.resolution);
    int gy = static_cast<int>(
      (pose.position.y - map_->info.origin.position.y) / map_->info.resolution);

    return GridPose(gx, gy);
  }

  geometry_msgs::msg::Pose gridToPose(const GridPose & gp) const
  {
    geometry_msgs::msg::Pose pose;
    pose.position.x =
      gp.x * map_->info.resolution + map_->info.origin.position.x;
    pose.position.y =
      gp.y * map_->info.resolution + map_->info.origin.position.y;
    pose.orientation.w = 1.0;
    return pose;
  }

  inline int gridToCell(const GridPose & gp) const
  {
    return gp.y * map_->info.width + gp.x;
  }

  bool lineCrossesObstacle(const std::vector<GridPose> & line) const
  {
    const auto & data = map_->data;

    for (const auto & gp : line) {
      int idx = gridToCell(gp);
      if (idx < 0 || idx >= static_cast<int>(data.size())) {
        return true;
      }
      if (data[idx] > cost_limit_) {
        return true;
      }
    }
    return false;
  }

  std::vector<GridPose> bresenhamLine(const GridPose & start,
                                     const GridPose & end) const
  {
    std::vector<GridPose> line;

    int x0 = start.x, y0 = start.y;
    int x1 = end.x,   y1 = end.y;

    int dx = std::abs(x1 - x0);
    int dy = std::abs(y1 - y0);

    int sx = (x0 < x1) ? 1 : -1;
    int sy = (y0 < y1) ? 1 : -1;

    int err = dx - dy;

    while (true) {
      line.emplace_back(x0, y0);

      if (x0 == x1 && y0 == y1)
        break;

      int e2 = 2 * err;
      if (e2 > -dy) {
        err -= dy;
        x0 += sx;
      }
      if (e2 < dx) {
        err += dx;
        y0 += sy;
      }
    }

    return line;
  }

  /* ===================== Members ===================== */

  rclcpp::Subscription<nav_msgs::msg::OccupancyGrid>::SharedPtr map_sub_;
  rclcpp::Subscription<nav_msgs::msg::Path>::SharedPtr path_sub_;
  rclcpp::Publisher<nav_msgs::msg::Path>::SharedPtr path_pub_;

  nav_msgs::msg::OccupancyGrid::SharedPtr map_;

  int iterations_{2};
  int cost_limit_{20};

  std::chrono::duration<double> testDuration;
  std::chrono::_V2::system_clock::time_point startTime;
  std::chrono::_V2::system_clock::time_point endTime;
};

/* ===================== main ===================== */

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<AStarSmoother>());
  rclcpp::shutdown();
  return 0;
}
